File: __tests__/lib/activity.test.ts
// Tests des métadonnées d'activité (lib/activity.ts) : la table qui décide
// partout ailleurs si une séance a un tracé GPS et si l'effort se lit en allure.
import {
  ACTIVITY_META,
  ACTIVITY_TYPES,
  isGpsActivity,
  toActivityType,
  usesPace,
} from '@/lib/activity';
import type { ActivityType } from '@/lib/types';

describe('ACTIVITY_META', () => {
  it('couvre exactement les types listés, sans doublon', () => {
    expect([...ACTIVITY_TYPES].sort()).toEqual(['course', 'marche', 'muscu', 'velo']);
    expect(new Set(ACTIVITY_TYPES).size).toBe(ACTIVITY_TYPES.length);
    expect(Object.keys(ACTIVITY_META).sort()).toEqual([...ACTIVITY_TYPES].sort());
  });

  it('donne à chaque activité un libellé, un libellé court, une icône et une teinte', () => {
    for (const type of ACTIVITY_TYPES) {
      const meta = ACTIVITY_META[type];
      expect(meta.label.length).toBeGreaterThan(0);
      expect(meta.shortLabel.length).toBeGreaterThan(0);
      expect(meta.icon.length).toBeGreaterThan(0);
      expect(meta.colorKey.length).toBeGreaterThan(0);
    }
  });

  it('donne une teinte distincte à chaque activité', () => {
    const keys = ACTIVITY_TYPES.map((t) => ACTIVITY_META[t].colorKey);
    expect(new Set(keys).size).toBe(keys.length);
  });
});

describe('isGpsActivity', () => {
  it('vrai pour le vélo, la course et la marche ; faux en musculation', () => {
    expect(isGpsActivity('velo')).toBe(true);
    expect(isGpsActivity('course')).toBe(true);
    expect(isGpsActivity('marche')).toBe(true);
    expect(isGpsActivity('muscu')).toBe(false);
  });
});

describe('usesPace', () => {
  it('allure à pied, vitesse à vélo', () => {
    expect(usesPace('course')).toBe(true);
    expect(usesPace('marche')).toBe(true);
    expect(usesPace('velo')).toBe(false);
    expect(usesPace('muscu')).toBe(false);
  });

  it('ne concerne que des activités tracées', () => {
    for (const type of ACTIVITY_TYPES) {
      if (usesPace(type)) expect(isGpsActivity(type)).toBe(true);
    }
  });
});

describe('toActivityType', () => {
  it('accepte les types connus', () => {
    for (const type of ACTIVITY_TYPES) expect(toActivityType(type)).toBe(type);
  });

  it('retombe sur le vélo pour une valeur inconnue ou absente', () => {
    expect(toActivityType(undefined)).toBe('velo');
    expect(toActivityType('natation')).toBe('velo');
    expect(toActivityType(42)).toBe('velo');
    // Une clé héritée d'Object.prototype ne doit pas passer pour un type.
    expect(toActivityType('toString')).toBe('velo');
  });

  it('honore le repli explicite', () => {
    const fallback: ActivityType = 'course';
    expect(toActivityType('inconnu', fallback)).toBe('course');
  });
});

File: app/src/test/java/ovh/battistella/elan/domain/ActivityTest.kt
// Tests des métadonnées d'activité : la table qui décide partout ailleurs si
// une séance a un tracé GPS et si l'effort se lit en allure.
package ovh.battistella.elan.domain

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class ActivityTest {
    @Test
    fun `ACTIVITY_META couvre exactement les types listés, sans doublon`() {
        assertEquals(listOf("course", "marche", "muscu", "velo"), ACTIVITY_TYPES.map { it.key }.sorted())
        assertEquals(ACTIVITY_TYPES.size, ACTIVITY_TYPES.toSet().size)
        assertEquals(ACTIVITY_TYPES.toSet(), ACTIVITY_META.keys)
        assertEquals(ActivityType.entries.toSet(), ACTIVITY_META.keys)
    }

    @Test
    fun `donne à chaque activité un libellé, un libellé court, une icône et une teinte`() {
        for (type in ACTIVITY_TYPES) {
            val meta = ACTIVITY_META.getValue(type)
            assertTrue(meta.label.isNotEmpty())
            assertTrue(meta.shortLabel.isNotEmpty())
            assertTrue(meta.icon.isNotEmpty())
            assertTrue(meta.colorKey.isNotEmpty())
        }
    }

    @Test
    fun `donne une teinte distincte à chaque activité`() {
        val keys = ACTIVITY_TYPES.map { ACTIVITY_META.getValue(it).colorKey }
        assertEquals(keys.size, keys.toSet().size)
    }

    @Test
    fun `isGpsActivity vrai pour le vélo, la course et la marche, faux en musculation`() {
        assertTrue(isGpsActivity(ActivityType.VELO))
        assertTrue(isGpsActivity(ActivityType.COURSE))
        assertTrue(isGpsActivity(ActivityType.MARCHE))
        assertFalse(isGpsActivity(ActivityType.MUSCU))
    }

    @Test
    fun `usesPace allure à pied, vitesse à vélo`() {
        assertTrue(usesPace(ActivityType.COURSE))
        assertTrue(usesPace(ActivityType.MARCHE))
        assertFalse(usesPace(ActivityType.VELO))
        assertFalse(usesPace(ActivityType.MUSCU))
    }

    @Test
    fun `usesPace ne concerne que des activités tracées`() {
        for (type in ACTIVITY_TYPES) {
            if (usesPace(type)) assertTrue(isGpsActivity(type))
        }
    }

    @Test
    fun `toActivityType accepte les types connus`() {
        for (type in ACTIVITY_TYPES) assertEquals(type, toActivityType(type.key))
    }

    @Test
    fun `toActivityType retombe sur le vélo pour une valeur inconnue ou absente`() {
        assertEquals(ActivityType.VELO, toActivityType(null))
        assertEquals(ActivityType.VELO, toActivityType("natation"))
        assertEquals(ActivityType.VELO, toActivityType(42))
        // Une clé héritée d'Object.prototype ne doit pas passer pour un type.
        assertEquals(ActivityType.VELO, toActivityType("toString"))
    }

    @Test
    fun `toActivityType honore le repli explicite`() {
        assertEquals(ActivityType.COURSE, toActivityType("inconnu", ActivityType.COURSE))
    }

    @Test
    fun `fromKey résout les clés stockées et rejette le reste`() {
        assertEquals(ActivityType.MARCHE, ActivityType.fromKey("marche"))
        assertEquals(null, ActivityType.fromKey("MARCHE"))
        assertEquals(null, ActivityType.fromKey(null))
        assertEquals(TrainingGoal.PERTE_POIDS, TrainingGoal.fromKey("perte-poids"))
        assertEquals(Difficulty.DUR, Difficulty.fromKey("dur"))
        assertEquals(Sex.F, Sex.fromKey("f"))
    }
}
